File: optimizers/RAdam.py
from .Optimizer import Optimizer, np


class RAdam(Optimizer):
    """
    RAdam — Adam с ректификацией дисперсии адаптивного шага (Rectified Adam).

    Первоисточник: Liu L., Jiang H., He P., Chen W., Liu X., Gao J., Han J.
    "On the Variance of the Adaptive Learning Rate and Beyond". arXiv:1908.03265, 2019 (ICLR 2020).
    https://arxiv.org/abs/1908.03265
    Реализация следует алгоритму из документации PyTorch:
    https://docs.pytorch.org/docs/stable/generated/torch.optim.RAdam.html
    """

    param_descriptions = {
        "lr": "скорость обучения — масштаб шага",
        "beta1": "коэффициент EMA первого момента (сглаживание направления градиента)",
        "beta2": "коэффициент EMA второго момента; от него же зависит порог включения ректификации",
        "eps": "малая добавка в знаменатель для численной устойчивости, настраивать обычно не нужно",
        "weight_decay": "L2-штраф: тянет точку к началу координат, искажая минимизируемую функцию",
    }

    v: np.ndarray
    acc: np.ndarray
    t: int

    def __init__(
        self,
        initial_x: np.ndarray,
        function,
        lr: float = 0.01,
        beta1: float = 0.9,
        beta2: float = 0.999,
        eps: float = 1e-08,
        weight_decay: float = 0.0,
    ) -> None:
        params = dict(lr=lr, beta1=beta1, beta2=beta2, eps=eps, weight_decay=weight_decay)
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)
        self.t += 1
        gradient = gradient + self.params["weight_decay"] * self.x
        self.v = self.params["beta1"] * self.v + (1 - self.params["beta1"]) * gradient
        self.acc = self.params["beta2"] * self.acc + (1 - self.params["beta2"]) * gradient**2
        bias_correction1 = 1 - self.params["beta1"] ** self.t
        bias_correction2 = 1 - self.params["beta2"] ** self.t
        corrected_v = self.v / bias_correction1
        rho_inf = 2 / (1 - self.params["beta2"]) - 1
        rho_t = rho_inf - 2 * self.t * self.params["beta2"] ** self.t / bias_correction2
        if rho_t > 5.0:
            rect = ((rho_t - 4) * (rho_t - 2) * rho_inf / ((rho_inf - 4) * (rho_inf - 2) * rho_t)) ** 0.5
            adaptive_lr = np.sqrt(bias_correction2) / (np.sqrt(self.acc) + self.params["eps"])
            next_x = self.x - self.params["lr"] * rect * adaptive_lr * corrected_v
        else:
            next_x = self.x - self.params["lr"] * corrected_v
        return self.move_next(next_x)

    def reset(self) -> None:
        super().reset()
        self.v = np.zeros([2])
        self.acc = np.zeros([2])
        self.t = 0

File: src/shared/lib/optimization-engine/optimizers/radam.ts
import type { OptimizerDescriptor } from "./types";
import { zero2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient } from "@shared/lib/optimization-engine/functions";

export const rAdamOptimizer: OptimizerDescriptor = {
  name: "RAdam",
  params: {
    lr: { default: 0.01, description: "скорость обучения — масштаб шага" },
    beta1: { default: 0.9, description: "коэффициент EMA первого момента (сглаживание направления градиента)" },
    beta2: {
      default: 0.999,
      description: "коэффициент EMA второго момента; от него же зависит порог включения ректификации",
    },
    eps: { default: 1e-8, description: "малая добавка в знаменатель для численной устойчивости, настраивать обычно не нужно" },
    weightDecay: {
      default: 0.0,
      description: "L2-штраф: тянет точку к началу координат, искажая минимизируемую функцию",
    },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let v: Vec2 = zero2();
    let acc: Vec2 = zero2();
    let t = 0;
    return {
      get x() { return x; },
      params,
      next() {
        const rawGrad = gradient(fn, x[0], x[1]);
        t += 1;
        const grad: Vec2 = [rawGrad[0] + params.weightDecay * x[0], rawGrad[1] + params.weightDecay * x[1]];
        v = [
          params.beta1 * v[0] + (1 - params.beta1) * grad[0],
          params.beta1 * v[1] + (1 - params.beta1) * grad[1],
        ];
        acc = [
          params.beta2 * acc[0] + (1 - params.beta2) * grad[0] ** 2,
          params.beta2 * acc[1] + (1 - params.beta2) * grad[1] ** 2,
        ];
        const biasCorrection1 = 1 - params.beta1 ** t;
        const biasCorrection2 = 1 - params.beta2 ** t;
        const correctedV: Vec2 = [v[0] / biasCorrection1, v[1] / biasCorrection1];
        const rhoInf = 2 / (1 - params.beta2) - 1;
        const rhoT = rhoInf - (2 * t * params.beta2 ** t) / biasCorrection2;
        if (rhoT > 5.0) {
          const rect = Math.sqrt(
            ((rhoT - 4) * (rhoT - 2) * rhoInf) / ((rhoInf - 4) * (rhoInf - 2) * rhoT),
          );
          const adaptiveLr: Vec2 = [
            Math.sqrt(biasCorrection2) / (Math.sqrt(acc[0]) + params.eps),
            Math.sqrt(biasCorrection2) / (Math.sqrt(acc[1]) + params.eps),
          ];
          x = [
            x[0] - params.lr * rect * adaptiveLr[0] * correctedV[0],
            x[1] - params.lr * rect * adaptiveLr[1] * correctedV[1],
          ];
        } else {
          x = [x[0] - params.lr * correctedV[0], x[1] - params.lr * correctedV[1]];
        }
        return { x, value: fn(x[0], x[1]), internals: { "v.x": v[0], "v.y": v[1], "|v|": Math.sqrt(v[0] ** 2 + v[1] ** 2), "acc.x": acc[0], "acc.y": acc[1], "|acc|": Math.sqrt(acc[0] ** 2 + acc[1] ** 2), t, rhoT } };
      },
      reset() {
        x = initialX;
        v = zero2();
        acc = zero2();
        t = 0;
      },
    };
  },
};
